### PR TITLE
fix: middleware to warn if response headers are invalid

### PR DESCRIPTION
**Pull request by dangdennis**
Resolves https://github.com/aantron/dream/issues/233. I was unable to replicate the socket hanging issue on alpine because I couldn't build it due to ssl dep issues. However, I ran the debian build on fly.io, and fly rejects http messages with invalid headers. So we should fix this regardless of linux distro. We can fix the alpine ssl issue separately.

- [x]  Learned that dropping empty headers via the builtin middleware will not work because it runs before the user's middleware, and definitely before the helper methods are called
- [x] Spiral in dune usage lol
- [x] Clean headers at lowest root response call

### CODE REVIEW

**Comment by aantron** [comment]
Thanks!

> * Learned that dropping empty headers via the builtin middleware will not work because it runs before the user's middleware, and definitely before the helper methods are called

Not yet a full review, but AFAIK built-in middleware runs both before and after the user's middleware:

```ocaml
let my_middleware user's_middleware = fun request ->
  do_stuff_before;
  let%lwt response = user's_middleware request in
  do_stuff_after;
  response
```

**Comment by dangdennis** [comment]
> Thanks!
> 
> > * Learned that dropping empty headers via the builtin middleware will not work because it runs before the user's middleware, and definitely before the helper methods are called
> 
> Not yet a full review, but AFAIK built-in middleware runs both before and after the user's middleware:
> 
> ```ocaml
> let my_middleware user's_middleware = fun request ->
>   do_stuff_before;
>   let%lwt response = user's_middleware request in
>   do_stuff_after;
>   response
> ```

Ah thanks for clarifying. I'll re-implement this as a built-in middleware built-in then to log a warning if there are empty header keys.

**Comment by dangdennis** [comment]
😆  

> AFAIK built-in middleware runs 

This cracks me up because I'm assuming you've written the middleware logic yourself.

**Comment by aantron** [comment]
> This cracks me up because I'm assuming you've written the middleware logic yourself.

Need to be careful in case I forgot or something has been broken since :)

**Comment by dangdennis** [comment]
unfortunately can't use expect-test against log outputs because we can't control the time when the logger executes. not that i can see any option. since i'm not performing any mutations anyways, think we're safe to not have a test.

File: test/expect/pure/message/message.ml [hunk 1]
```diff
@@ -3,9 +3,7 @@
 
    Copyright 2022 Anton Bachin *)
 
-
-
-let%expect_test _ =
+let%expect_test "middleware runs sequentially onion-style" =
```
Comment (dangdennis) [inline test/expect/pure/message/message.ml:6]: autoformatter kicked in for the file too.

**Comment by aantron** [comment]
I can merge this now to avoid extending this PR unnecessarily. However, as I recall, I wanted to move Dream away from the built-in middleware model. That's why there was only one left, and the plan was to find a way to get rid of it as a built-in middleware as well. I think this warning can go unconditionally into the HTTP adapters -- hopefully in a place where it will be triggered for both HTTP1 and HTTP2. This area:

https://github.com/aantron/dream/blob/a2961cd7fa150903db758fa3737213cddd7ddb3b/src/http/http.ml#L407-L448

already has several wrappers around the user's handler, so we just need to add the middleware here unconditionally. It should run outside of the `catch` middleware, since `catch` might also somehow set an empty header.

**Comment by dangdennis** [comment]
I'd prefer do this as a follow-up too - re: move the header check to h1/h2 paths. I thought it was a little strange to have important checks be opt-in, so I'm glad to see this direction change. The gold standard imo is asp.net & phoenix. Baking in security and best-practices seems best.

File: test/expect/http/dune [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+ (name test_expect_http)
+ (libraries
+  base
+  dream
+  dream.http
+  lwt
+  lwt.unix
+  ppx_expect.common
+  ppx_inline_test.config
+  ppx_expect.config_types))
```
Comment (dangdennis) [inline test/expect/http/dune:11]: leaving these dune files here for the future instead of having to set them up again.

File: src/http/http.ml [hunk 3]
```diff
@@ -380,15 +380,20 @@ let ocaml_tls = {
...
+  let invalid_headers_exist = Message.all_headers response |> List.exists (fun (key, _) -> String.trim key = "") in
+  if invalid_headers_exist then
+    log.warning (fun log ->
+      log "A response header is empty or contains only whitespace. This is not allowed by RFC 7230.");
+  Lwt.return response
 
 let built_in_middleware error_handler =
   Message.pipeline [
     Catch.catch (Error_handler.app error_handler);
+    check_headers_middleware;
```
Comment (aantron) [inline src/http/http.ml:394]: I think this should be the outer middleware, as the error handler can also be a source of empty headers, while, conversely, `check_headers_middleware` cannot raise any exceptions.

**Comment by aantron** [comment]
> I'd prefer do this as a follow-up too - re: move the header check to h1/h2 paths. I thought it was a little strange to have important checks be opt-in, so I'm glad to see this direction change. The gold standard imo is asp.net & phoenix. Baking in security and best-practices seems best.

Not sure what you mean here -- built-in middlewares are/were opt-out, they are included in the middleware stack by default.

**Comment by dangdennis** [comment]
> > I'd prefer do this as a follow-up too - re: move the header check to h1/h2 paths. I thought it was a little strange to have important checks be opt-in, so I'm glad to see this direction change. The gold standard imo is asp.net & phoenix. Baking in security and best-practices seems best.
> 
> Not sure what you mean here -- built-in middlewares are/were opt-out, they are included in the middleware stack by default.

Oops! Yes you're right. It's already defaulted to true. I forgot to check the actual default value 😅 . I only ever read the lines below, and not specifically the args set on `run`.

```
  let user's_dream_handler =
    if builtins then
      built_in_middleware error_handler user's_dream_handler
    else
      user's_dream_handler
  in
```

**Comment by aantron** [comment]
I've made a few edits. Notably:

```diff
diff --git a/src/http/http.ml b/src/http/http.ml
index 11943be..d1b7614 100644
--- a/src/http/http.ml
+++ b/src/http/http.ml
@@ -390,7 +390,7 @@ let check_headers_middleware next_handler request =
   in
   if invalid_headers_exist then
     log.warning (fun log ->
-      log "A response header is empty or contains only whitespace");
+      log ~request "A response header is empty or contains only whitespace");
   Lwt.return response

 let built_in_middleware error_handler =
```

changes the log output from

```
23.04.23 14:13:42.184      dream.http  WARN A response header is empty or contains only whitespace
```

to

```
23.04.23 14:13:42.184      dream.http  WARN REQ 1 A response header is empty or contains only whitespace
```

I then tested example `9-error` to see if this PR would warn about empty headers set in the error handler:

```diff
diff --git a/example/9-error/error.eml.ml b/example/9-error/error.eml.ml
index 515091c..8a995d7 100644
--- a/example/9-error/error.eml.ml
+++ b/example/9-error/error.eml.ml
@@ -12,6 +12,7 @@ let my_error_template _error debug_info suggested_response =
     </body>
     </html>
   end;
+  Dream.set_header suggested_response "" "abc";
   Lwt.return suggested_response

 let () =
```

The PR didn't warn about this header, as expected, so I added

```diff
diff --git a/src/http/http.ml b/src/http/http.ml
index d1b7614..2deb1db 100644
--- a/src/http/http.ml
+++ b/src/http/http.ml
@@ -395,8 +395,8 @@ let check_headers_middleware next_handler request =

 let built_in_middleware error_handler =
   Message.pipeline [
-    Catch.catch (Error_handler.app error_handler);
     check_headers_middleware;
+    Catch.catch (Error_handler.app error_handler);
   ]


```

It now warns about empty headers set in error handlers:

```
23.04.23 14:17:19.285       dream.log  INFO REQ 1 GET / 127.0.0.1:39224 Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36
23.04.23 14:17:19.285       dream.log  WARN REQ 1 404 in 62 μs
23.04.23 14:17:19.285      dream.http  WARN REQ 1 A response header is empty or contains only whitespace
```

**Comment by aantron** [comment]
Thank you! As for built-in middleware, it's probably fine to leave the implementation in `http.ml` as it is now. I'll probably just hide them (their remainder, `Dream.catch`) from `dream.mli`. All the other former built-in middlewares have already been factored away anyway, and all that's left is a few that we want to keep around unconditionally, and it's fine to internally interpret them as a middleware stack.

**Pull request merged**